sympy/geometry/ellipse.py

```diff
@@ -1339,30 +1341,81 @@ class Circle(Ellipse):
     >>> c2.hradius, c2.vradius, c2.radius, c2.center
     (sqrt(2)/2, sqrt(2)/2, sqrt(2)/2, Point2D(1/2, 1/2))
 
+    >>> # a circle object returned from given equation and optional 'x' and 'y' parameters
+    >>> from sympy.abc import x, y, a, b
+    >>> Circle(x ** 2 + y ** 2 - 25)
+    Circle(Point2D(0, 0), 5)
+    >>> Circle(a ** 2 + b ** 2 - 25, x='a', y='b')
+    Circle(Point2D(0, 0), 5)
+
     """
 
     def __new__(cls, *args, **kwargs):
-        c, r = None, None
-        if len(args) == 3:
-            args = [Point(a, dim=2) for a in args]
-            if Point.is_collinear(*args):
-                raise GeometryError(
-                    "Cannot construct a circle from three collinear points")
-            from .polygon import Triangle
-            t = Triangle(*args)
-            c = t.circumcenter
-            r = t.circumradius
-        elif len(args) == 2:
-            # Assume (center, radius) pair
-            c = Point(args[0], dim=2)
-            r = sympify(args[1])
-
-        if not (c is None or r is None):
-            if r == 0:
-                return c
-            return GeometryEntity.__new__(cls, c, r, **kwargs)
-
-        raise GeometryError("Circle.__new__ received unknown arguments")
+
+        if len(args) == 1 and isinstance(args[0], Expr):
+            x = kwargs.get('x', 'x')
+            y = kwargs.get('y', 'y')
+            equation = args[0]
+
+            def find(x, equation):
+                free = equation.free_symbols
+                xs = [i for i in free if (i.name if type(x) is str else i) == x]
+                if not xs:
+                    raise ValueError('could not find %s' % x)
+                if len(xs) != 1:
+                    raise ValueError('ambiguous %s' % x)
+                return xs[0]
+
+            x = find(x, equation)
+            y = find(y, equation)
+
+            try:
```

Comment:
If you want to try this with linear_coeffs, consider
```python
 x2 = Dummy('%s**2' % x.name)
 y2 = Dummy('%s**2' % y.name)
 eq = _mexpand(equation, recursive=True).xreplace({x**2: x2, y**2: y2})
 co = linear_coeffs(eq, x2, y2, x, y)
 a, b, c, d, e = [co[i] for i in (x2, y2, x, y, 0)]
 if a == 0 or b == 0 or a != b:
     not a circle
 center_x = -c/a/2
 center_y = -d/b/2
 r2 = (center_x**2) + (center_y**2) - e
 if r2.is_negative:
     not a real circle 
 return Circle((center_x, center_y), sqrt(r2))
```

Reply:
@smichr  Yeah I was thinking about the same 
although I think these
```
x2 = Dummy('%s**2' % x.name)
y2 = Dummy('%s**2' % y.name)
eq = _mexpand(equation, recursive=True).xreplace({x**2: x2, y**2: y2})
```
would be very useful as they will fail for `Circle(a ** 2 + b ** 2 - 25, x='a', y='b')` ( I tried and they do fail )

So for their replacement, let's keep the `find` function as it's able to handle inputs in different formats well.

I've made another commit. Please take a look.